hypervisor/dm/vpci/msi.c

```diff
@@ -0,0 +1,230 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NETAPP, INC ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL NETAPP, INC OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * $FreeBSD$
+ */
+
+#include <hypervisor.h>
+#include "pci_priv.h"
+
+static inline bool msicap_access(struct pci_vdev *vdev, uint32_t offset)
+{
+	if (vdev->msi.capoff == 0U) {
+		return 0;
+	}
+
+	return in_range(offset, vdev->msi.capoff, vdev->msi.caplen);
+}
+
+static int vmsi_remap(struct pci_vdev *vdev, bool enable)
+{
+	struct ptdev_msi_info info;
+	union pci_bdf pbdf = vdev->pdev.bdf;
+	struct vm *vm = vdev->vpci->vm;
+	uint32_t capoff = vdev->msi.capoff;
+	uint16_t msgctrl, msgdata;
+	uint32_t addrlo, addrhi;
+	int ret;
+
+	/* Disable MSI during configuration */
+	msgctrl = pci_vdev_read_cfg(vdev, capoff + PCIR_MSI_CTRL, 2U);
+	if ((msgctrl & PCIM_MSICTRL_MSI_ENABLE) == PCIM_MSICTRL_MSI_ENABLE) {
+		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_CTRL, 2U, msgctrl & ~PCIM_MSICTRL_MSI_ENABLE);
+	}
+
+	/* Read the MSI capability structure from virtual device */
+	addrlo = pci_vdev_read_cfg_u32(vdev, capoff + PCIR_MSI_ADDR);
+	if (msgctrl & PCIM_MSICTRL_64BIT) {
+		msgdata = pci_vdev_read_cfg_u16(vdev, capoff + PCIR_MSI_DATA_64BIT);
+		addrhi = pci_vdev_read_cfg_u32(vdev, capoff + PCIR_MSI_ADDR_HIGH);
+	} else {
+		msgdata = pci_vdev_read_cfg_u16(vdev, capoff + PCIR_MSI_DATA);
+		addrhi = 0U;
+	}
+
+	info.is_msix = 0;
+	info.vmsi_addr = addrlo | ((uint64_t)addrhi << 32U);
+
+	/* MSI is being enabled or disabled */
+	if (enable) {
+		info.vmsi_data = msgdata;
+	} else {
+		info.vmsi_data = 0;
+	}
+
+	ret = ptdev_msix_remap(vm, vdev->vbdf.value, 0U, &info);
+	if (ret != 0U) {
+		return ret;
+	}
+
+	/* Update MSI Capability structure to physical device */
+	pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_ADDR, 0x4U, info.pmsi_addr);
+	if (msgctrl & PCIM_MSICTRL_64BIT) {
+		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_ADDR_HIGH, 0x4U,	info.pmsi_addr >> 32U);
+		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_DATA_64BIT, 0x2U, info.pmsi_data);
+	} else {
+		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_DATA, 0x2U, info.pmsi_data);
+	}
+
+	/* If MSI Enable is being set, set INTxDIS bit, otherwise clear it */
+	enable_disable_pci_intx(pbdf, !enable);
```

Comment:
It seems that this function will be called frequently for the given pci device.
(disable or enable intx)
In fact when MSI is supported, we seldom reenable the Intx again.
As it is called in vmsi_remap, it can always pass the 0.

Reply:
Yes, just need to make sure it's disabled.